File: src/hawkeye/extract.py
"""元素文本提取（纯粹围绕 Playwright Page 的读取与归一化）。

放在独立模块，便于用 ``page.set_content`` 加载本地 HTML 夹具直接测试，
无需真实网络导航。除单元素文本外，另提供列表页的「标题链接」批量提取。
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from urllib.parse import urljoin, urlsplit

from playwright.async_api import Error as PlaywrightError
from playwright.async_api import Page
from playwright.async_api import TimeoutError as PlaywrightTimeoutError

from .config import MonitoredElement, WatchTarget

logger = logging.getLogger(__name__)

_WS = re.compile(r"\s+")


def normalize_text(s: str) -> str:
    """去首尾空白，并把内部连续空白（含换行/缩进）折叠为单个空格。"""
    return _WS.sub(" ", s).strip()


@dataclass(frozen=True)
class ExtractResult:
    """元素提取结果。

    ``value`` = 归一化后的字符串（或空串、None）。
    ``reason`` = 失败原因（仅在 ``value`` 不可用时非 None）；成功时为 None。
    文本模式与 JS 模式共用同一套语义，区别仅在 reason 文案——fetch.py 不再需要分支 if/elif，
    直接把 reason 转给 FetchNoMatch，由 scheduler / Telegram 失败告警原样转达。
    """

    value: str | None
    reason: str | None = None


def _prefixed_selector(selector: str, effective_type: str) -> str:
    """把配置选择器转成 Playwright 明确前缀形式，避免 auto 猜测歧义。"""
    sel = selector.strip()
    if effective_type == "xpath":
        return sel if sel.startswith("xpath=") else f"xpath={sel}"
    return sel


def _playwright_selector(element: MonitoredElement) -> str:
    """单元素选择器的明确前缀形式。"""
    return _prefixed_selector(element.selector, element.effective_selector_type)


async def extract_text(page: Page, element: MonitoredElement, timeout_ms: int) -> ExtractResult:
    """取匹配元素的状态值，连同失败原因一起返回。

    文本模式（``element.js`` 缺省）：等元素 attached，取 inner_text（空时 text_content 兜底），
    经 normalize_text 归一化。失败时返回 ``(None, reason)`` 或 ``("", reason)``，fetch.py
    据此区分「未匹配」与「匹配但值为空」。

    JS 模式（``element.js`` 非空）：先等元素 attached，再 ``loc.element_handle()`` 拿真实
    DOMElement 句柄交给 ``page.evaluate(js, handle)``。element handle stale 时
    page.evaluate 会抛 PlaywrightError（走 fetch.py 的现有 except 归到 FetchError），
    而非 ``loc.evaluate`` 在某些场景下静默返回 undefined 后归到「JS 表达式未返回值」
    （掩盖真实原因）。

    所有失败路径都返回 reason：文本模式是「选择器未匹配到元素」/「选择器匹配到元素，但其
    文本为空」；JS 模式根据失败点给出更具体的文案（含 handle / evaluate / 选择器未匹配）。
    """
    idx = element.nth if element.nth is not None else 0
    loc = page.locator(_playwright_selector(element)).nth(idx)

    try:
        # 导航只等到 domcontentloaded，JS 渲染的元素此刻可能还没挂载，必须在这里等。
        # 等 attached 而非 visible：隐藏元素也要能取到，靠下面 text_content 兜底。
        await loc.wait_for(state="attached", timeout=timeout_ms)
    except PlaywrightTimeoutError:
        # 挑战页是页面级现象,与 JS / 文本模式无关——一律把页面 title 塞到 reason,
        # 让用户从 Telegram 失败告警直接看出「被反爬拦了」（典型 title「Just a moment...」）,
        # 而不是干巴巴的「选择器未匹配」。取 title 失败时退回到纯静态文案,不阻断失败路径。
        try:
            title = await page.title()
        except PlaywrightError:
            title = ""
        reason = "选择器未匹配到元素"
        if title:
            reason = f"选择器未匹配到元素（页面标题：{title!r}，可能是反爬挑战页）"
        return ExtractResult(value=None, reason=reason)

    if element.js is not None:
        handle = await loc.element_handle()
        if handle is None:
            return ExtractResult(
                value=None,
                reason="JS 模式 element_handle 不可用（strict mode 违规或元素被移除）",
            )
        try:
            raw = await page.evaluate(element.js, handle)
        except PlaywrightError as e:
            return ExtractResult(value=None, reason=f"JS 执行失败：{_first_line(str(e))}")
        if raw is None:
            return ExtractResult(value=None, reason="JS 表达式未返回值")
        return ExtractResult(value=normalize_text(str(raw)), reason=None)

    raw = await loc.inner_text()
    if not raw.strip():
        raw = await loc.text_content() or ""
    if not raw:
        return ExtractResult(
            value="",
            reason="选择器匹配到元素，但其文本为空（多半指向了纯装饰节点，试试上一级）",
        )
    return ExtractResult(value=normalize_text(raw), reason=None)


def _first_line(s: str) -> str:
    """取第一行：Playwright 异常消息常含多行堆栈，嵌入 reason 前先压扁。"""
    return s.splitlines()[0] if s else s


# ---- 列表页标题链接提取 ----


@dataclass(frozen=True)
class ListItem:
    """列表页中的单个帖子项：帖子 ID、标题、可点击的绝对 URL。"""

    post_id: str
    title: str
    url: str


def _extract_id(href: str, pattern: re.Pattern[str] | None) -> str | None:
    """从 href 提取帖子 ID：有 id_pattern 用其捕获组 1，否则取 URL path 末段。"""
    if pattern is not None:
        m = pattern.search(href)
        # 正则可能匹配但无捕获组（如 id_pattern 写成无分组）——此时无从取 ID，跳过。
        if m is None or not m.groups():
            return None
        return m.group(1)
    path = urlsplit(href).path.rstrip("/")
    if not path:
        return None
    return path.rsplit("/", 1)[-1] or None


async def extract_list_items(page: Page, watch: WatchTarget, timeout_ms: int) -> list[ListItem]:
    """提取列表页每个帖子的 (ID, 标题, 绝对 URL)。

    用 ``watch.link_selector`` 定位全部标题链接 ``<a>``：逐个读 href 与文本，href 经
    ``id_pattern``（缺省取 path 末段）提取帖子 ID、经 ``urljoin`` 补全为绝对 URL，
    文本经 :func:`normalize_text` 折叠空白作标题。缺 href 或提取不到 ID 的链接跳过
    （debug 日志、绝不抛异常）；选择器无匹配时返回空列表，交由上层归入失败路径。
    """
    loc = page.locator(_prefixed_selector(watch.link_selector, watch.effective_selector_type))

    try:
        # 与 extract_text 同理：domcontentloaded 落地时列表可能尚未渲染，先等首项挂载。
        await loc.first.wait_for(state="attached", timeout=timeout_ms)
    except PlaywrightTimeoutError:
        return []

    pattern = re.compile(watch.id_pattern) if watch.id_pattern is not None else None
    items: list[ListItem] = []
    for i in range(await loc.count()):
        a = loc.nth(i)
        href = await a.get_attribute("href")
        if not href:
            logger.debug("列表项缺少 href，跳过：watch=%s 第 %d 项", watch.name, i)
            continue
        post_id = _extract_id(href, pattern)
        if post_id is None:
            logger.debug("列表项无法提取帖子 ID，跳过：watch=%s href=%s", watch.name, href)
            continue
        raw_title = await a.inner_text()
        if not raw_title.strip():
            raw_title = await a.text_content() or ""
        items.append(
            ListItem(post_id=post_id, title=normalize_text(raw_title), url=urljoin(watch.url, href))
        )
    return items

File: rust/src/extract.rs
//! 元素文本提取（纯粹围绕 CDP 页面的读取与归一化）。
//!
//! Playwright 语义在 chromiumoxide 上的等价实现：CSS 选择器走 querySelector，
//! XPath 走 document.evaluate；「等元素 attached」用带超时的轮询 evaluate 实现。

use std::time::Duration;

use chromiumoxide::Page;

use crate::config::{MonitoredElement, WatchTarget};

/// 去首尾空白，并把内部连续空白（含换行/缩进）折叠为单个空格。
pub fn normalize_text(s: &str) -> String {
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// 元素提取结果：value = 归一化后的字符串（或空串、None）；
/// reason = 失败原因（仅在 value 不可用时非 None）。
#[derive(Debug, Clone, PartialEq)]
pub struct ExtractResult {
    pub value: Option<String>,
    pub reason: Option<String>,
}

/// 等元素挂载（attached）：在超时窗口内以 100ms 间隔轮询，可被提前打断。
async fn wait_attached(page: &Page, selector_js: &str, timeout: Duration) -> bool {
    let deadline = tokio::time::Instant::now() + timeout;
    let expr =
        format!("(()=>{{const els={selector_js};return els&&els.length>0?els[0]:els||null}})()");
    loop {
        let found = page
            .evaluate_expression(expr.clone())
            .await
            .ok()
            .and_then(|r| r.into_value::<serde_json::Value>().ok())
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        if found {
            return true;
        }
        if tokio::time::Instant::now() >= deadline {
            return false;
        }
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
}

/// 等所有匹配挂载后的全部元素数组表达式。
fn all_selector_expr(selector: &str, effective_type: &str) -> String {
    let sel = selector.trim();
    if effective_type == "xpath" {
        let body = sel.strip_prefix("xpath=").unwrap_or(sel);
        format!(
            "(()=>{{const r=document.evaluate({body:?},document,null,XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,null);const out=[];for(let i=0;i<r.snapshotLength;i++)out.push(r.snapshotItem(i));return out}})()"
        )
    } else {
        let body = sel.strip_prefix("css=").unwrap_or(sel);
        format!("Array.from(document.querySelectorAll({body:?}))")
    }
}

/// 谓词表达式：匹配集是否非空（用于首项挂载等待）。
fn any_selector_expr(selector: &str, effective_type: &str) -> String {
    let sel = selector.trim();
    if effective_type == "xpath" {
        let body = sel.strip_prefix("xpath=").unwrap_or(sel);
        format!(
            "(()=>{{const r=document.evaluate({body:?},document,null,XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,null);return r.snapshotLength>0}})()"
        )
    } else {
        let body = sel.strip_prefix("css=").unwrap_or(sel);
        format!("document.querySelector({body:?})!==null")
    }
}

async fn eval_string(page: &Page, expr: &str) -> Option<String> {
    page.evaluate_expression(expr.to_string())
        .await
        .ok()
        .and_then(|r| r.into_value::<serde_json::Value>().ok())
        .and_then(|v| match v {
            serde_json::Value::String(s) => Some(s),
            serde_json::Value::Null => None,
            other => Some(other.to_string()),
        })
}

/// 取匹配元素的状态值，连同失败原因一起返回。
///
/// 文本模式（element.js 缺省）：等元素 attached，取 innerText（空时 textContent 兜底），
/// 经 normalize_text 归一化。JS 模式：先等元素 attached，再对元素求值 JS 表达式。
pub async fn extract_text(
    page: &Page,
    element: &MonitoredElement,
    timeout: Duration,
) -> ExtractResult {
    let idx = element.nth.unwrap_or(0);
    let sel_expr = all_selector_expr(&element.selector, element.effective_selector_type());

    // 等首项挂载（domcontentloaded 后 JS 渲染元素可能还没挂载）。
    if !wait_attached(
        page,
        &any_selector_expr(&element.selector, element.effective_selector_type()),
        timeout,
    )
    .await
    {
        let title = page.get_title().await.ok().flatten().unwrap_or_default();
        let reason = if title.is_empty() {
            "选择器未匹配到元素".to_string()
        } else {
            format!("选择器未匹配到元素（页面标题：{title:?}，可能是反爬挑战页）")
        };
        return ExtractResult {
            value: None,
            reason: Some(reason),
        };
    }

    let nth_expr =
        format!("(()=>{{const els={sel_expr};return els.length>{idx}?els[{idx}]:null}})()");

    if let Some(js) = &element.js {
        // JS 模式：把元素作为参数求值。表达式的 this 是元素；兼容 (el) => ... 形式。
        let wrapped = format!(
            "(()=>{{const el={nth_expr};if(!el)return null;const fn={js};return typeof fn==='function'?fn.call(el,el):fn}})()"
        );
        match page.evaluate_expression(wrapped).await {
            Ok(r) => match r.into_value::<serde_json::Value>() {
                Ok(serde_json::Value::Null) => {
                    return ExtractResult {
                        value: None,
                        reason: Some("JS 表达式未返回值".to_string()),
                    };
                }
                Ok(v) => {
                    let s = match v {
                        serde_json::Value::String(s) => s,
                        other => other.to_string(),
                    };
                    return ExtractResult {
                        value: Some(normalize_text(&s)),
                        reason: None,
                    };
                }
                Err(_) => {
                    return ExtractResult {
                        value: None,
                        reason: Some("JS 执行结果不可读".to_string()),
                    };
                }
            },
            Err(e) => {
                return ExtractResult {
                    value: None,
                    reason: Some(format!("JS 执行失败：{e}")),
                };
            }
        }
    }

    // 文本模式：innerText，空则 textContent 兜底。
    let text_expr = format!(
        "(()=>{{const el={nth_expr};if(!el)return null;const it=el.innerText||'';const t=it.trim()?it:(el.textContent||'');return t}})()"
    );
    match eval_string(page, &text_expr).await {
        None => ExtractResult {
            value: None,
            reason: Some("元素句柄不可用".to_string()),
        },
        Some(t) if t.is_empty() => ExtractResult {
            value: Some(String::new()),
            reason: Some(
                "选择器匹配到元素，但其文本为空（多半指向了纯装饰节点，试试上一级）".to_string(),
            ),
        },
        Some(t) => ExtractResult {
            value: Some(normalize_text(&t)),
            reason: None,
        },
    }
}

/// 列表页中的单个帖子项：帖子 ID、标题、可点击的绝对 URL。
#[derive(Debug, Clone, PartialEq)]
pub struct ListItem {
    pub post_id: String,
    pub title: String,
    pub url: String,
}

pub(crate) fn extract_id(href: &str, pattern: Option<&regex::Regex>) -> Option<String> {
    if let Some(pattern) = pattern {
        let m = pattern.captures(href)?;
        return m.get(1).map(|g| g.as_str().to_string());
    }
    let path = url::Url::parse(href)
        .map(|u| u.path().trim_end_matches('/').to_string())
        .unwrap_or_else(|_| href.trim_end_matches('/').to_string());
    if path.is_empty() {
        return None;
    }
    path.rsplit('/')
        .next()
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
}

/// 提取列表页每个帖子的 (ID, 标题, 绝对 URL)。
///
/// 缺 href 或提取不到 ID 的链接跳过；选择器无匹配时返回空列表，交由上层归入失败路径。
pub async fn extract_list_items(
    page: &Page,
    watch: &WatchTarget,
    timeout: Duration,
) -> Vec<ListItem> {
    let any_expr = any_selector_expr(&watch.link_selector, watch.effective_selector_type());
    if !wait_attached(page, &any_expr, timeout).await {
        return Vec::new();
    }

    let pattern = watch
        .id_pattern
        .as_deref()
        .and_then(|p| regex::Regex::new(p).ok());
    let base = url::Url::parse(&watch.url).ok();

    let items_expr = all_selector_expr(&watch.link_selector, watch.effective_selector_type());
    // 一次性在页面内展开全部 (href, text) 对，避免逐项 round-trip。
    let expr = format!(
        "(()=>{{const els={items_expr};return els.map(a=>({{href:a.getAttribute('href')||'',text:(a.innerText&&a.innerText.trim())?a.innerText:(a.textContent||'')}}))}})()"
    );
    let raw: Vec<serde_json::Value> = match page.evaluate_expression(expr).await {
        Ok(r) => r.into_value().unwrap_or_default(),
        Err(_) => return Vec::new(),
    };

    let mut items = Vec::new();
    for entry in raw {
        let href = entry.get("href").and_then(|v| v.as_str()).unwrap_or("");
        let text = entry.get("text").and_then(|v| v.as_str()).unwrap_or("");
        if href.is_empty() {
            tracing::debug!("列表项缺少 href，跳过：watch={}", watch.name);
            continue;
        }
        let Some(post_id) = extract_id(href, pattern.as_ref()) else {
            tracing::debug!(
                "列表项无法提取帖子 ID，跳过：watch={} href={}",
                watch.name,
                href
            );
            continue;
        };
        let abs_url = match &base {
            Some(b) => match b.join(href) {
                Ok(joined) => joined.to_string(),
                Err(_) => href.to_string(),
            },
            None => href.to_string(),
        };
        items.push(ListItem {
            post_id,
            title: normalize_text(text),
            url: abs_url,
        });
    }
    items
}

/// 把相对 URL 补全为绝对 URL（公开给测试用）。
pub fn urljoin(base: &str, href: &str) -> String {
    match url::Url::parse(base).ok().and_then(|b| b.join(href).ok()) {
        Some(joined) => joined.to_string(),
        None => href.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_text() {
        assert_eq!(normalize_text("  a \n\t b   c "), "a b c");
        assert_eq!(normalize_text("充分\n  空白"), "充分 空白");
        assert_eq!(normalize_text(""), "");
    }

    #[test]
    fn test_extract_id_default_path_tail() {
        assert_eq!(
            extract_id("/post-911200-1", None),
            Some("post-911200-1".to_string())
        );
        assert_eq!(extract_id("/a/b/123/", None), Some("123".to_string()));
        assert_eq!(extract_id("", None), None);
        assert_eq!(extract_id("/", None), None);
    }

    #[test]
    fn test_extract_id_pattern_captures_group_one() {
        let re = regex::Regex::new(r"post-(\d+)-").unwrap();
        assert_eq!(
            extract_id("/post-911200-1", Some(&re)),
            Some("911200".to_string())
        );
    }

    #[test]
    fn test_extract_id_pattern_without_group_returns_none() {
        let re = regex::Regex::new(r"post-\d+-").unwrap();
        assert_eq!(extract_id("/post-911200-1", Some(&re)), None);
    }

    #[test]
    fn test_urljoin() {
        assert_eq!(urljoin("https://e.com/list", "/a/b"), "https://e.com/a/b");
        assert_eq!(
            urljoin(
                "https://www.nodeseek.com/",
                "https://www.nodeseek.com/post-1-1"
            ),
            "https://www.nodeseek.com/post-1-1"
        );
    }
}
